language/tools/move-build/src/source_package/manifest_parser.rs

```diff
@@ -0,0 +1,447 @@
+const DEV_ADDRESSES_NAME: &str = "dev_addresses";
+const DEPENDENCY_NAME: &str = "dependencies";
+const DEV_DEPENDENCY_NAME: &str = "dev_dependencies";
+
+const KNOWN_NAMES: &[&str] = &[
+    PACKAGE_NAME,
+    BUILD_NAME,
+    ADDRESSES_NAME,
+    DEV_ADDRESSES_NAME,
+    DEPENDENCY_NAME,
+    DEV_DEPENDENCY_NAME,
+];
+
+const REQUIRED_FIELDS: &[&str] = &[PACKAGE_NAME];
+
+pub fn parse_toml_and_unwrap_errors(manifest_string: String) -> Result<TV> {
+    match toml::from_str::<TV>(&manifest_string) {
+        Ok(toml_val) => Ok(toml_val),
+        Err(err) => Err(format_err!(
+            "Unable to parse as valid TOML.\nParsing error is:\n{}",
+            err
+        )),
+    }
+}
+
+pub fn parse_source_manifest(tval: TV) -> Result<PM::SourceManifest> {
+    match tval {
+        TV::Table(mut table) => {
+            check_for_required_field_names(&table, REQUIRED_FIELDS)
+                .context("Error parsing package manifest")?;
+            check_for_unknown_field_names(&table, KNOWN_NAMES)
+                .context("Error parsing package manifest")?;
+            let addresses = table
+                .remove(ADDRESSES_NAME)
+                .map(parse_addresses)
+                .transpose()
+                .context("Error parsing '[addresses]' section of manifest")?;
+            let dev_address_assignments = table
+                .remove(DEV_ADDRESSES_NAME)
+                .map(parse_dev_addresses)
+                .transpose()
+                .context("Error parsing '[dev_addresses]' section of manifest")?;
+            let package = table
+                .remove(PACKAGE_NAME)
+                .map(parse_package_info)
+                .transpose()
+                .context("Error parsing '[package]' section of manifest")?
+                .unwrap();
+            let build = table
+                .remove(BUILD_NAME)
+                .map(parse_build_info)
+                .transpose()
+                .context("Error parsing '[build]' section of manifest")?;
+            let dependencies = table
+                .remove(DEPENDENCY_NAME)
+                .map(parse_dependencies)
+                .transpose()
+                .context("Error parsing '[dependencies]' section of manifest")?
+                .unwrap_or_else(BTreeMap::new);
+            let dev_dependencies = table
+                .remove(DEV_DEPENDENCY_NAME)
+                .map(parse_dependencies)
+                .transpose()
+                .context("Error parsing '[dev_dependencies]' section of manifest")?
+                .unwrap_or_else(BTreeMap::new);
+            Ok(PM::SourceManifest {
+                package,
+                addresses,
+                dev_address_assignments,
+                build,
+                dependencies,
+                dev_dependencies,
+            })
+        }
+        x => {
+            bail!(
+                "Ill-formed package manifest {}. Expected a table at top level, but encountered a {}",
```

Comment:
How about "malformed" instead?

Reply:
There are a number of other instances of this in the PR, so whatever you do, please make them consistent.

Reply:
I like "malformed." I'll update throughout.